all/CMakeLists.txt

```diff
@@ -0,0 +1,57 @@
+# This file should be used for development with IDEs for the best discovery
+cmake_minimum_required(VERSION 3.8)
+
+project(Result_all CXX)
+
+set(Result_INCLUDE_WITHOUT_SYSTEM YES)
+set(CMAKE_INSTALL_INCLUDEDIR include/Result)
+
+add_subdirectory(.. sub/root)
```

Comment:
> Having CMakeLists.txt that add_subdirectory outside of a subdirectory is not a clean or modern CMake structure

It **is** in fact precisely modern CMake, as this is a superbuild CML that spares the project CMLs from the nasty details. The arguments to `add_subdirectory` were never required to be subdirectories of the current directory, its name is this for historic reasons. This command merely reads the CML from the given source directory and executes it in a child scope to the current scope.

Reply:
> The arguments to `add_subdirectory` were never required to be subdirectories of the current directory, its name is this for historic reasons

None of my concerns with this approach come from the name `add_subdirectory`, it comes from years of architectural design where project topology fundamentally matters.

> It is in fact precisely modern CMake, as this is a superbuild CML that **spares the project CMLs from the nasty details.** 

I'd be interested to see sources on why this is better or "modern", if you happen to have any. Suggesting that this is somehow better is completely foreign to me, and from what I can tell, these same benefits can be better achieved in many ways without having to resort to violating the topology of the structure.

Idiomatic CMake generally matches the topology of the project structure because it helps to convey the hierarchical nature of all dependent projects in a logical way. Using `add_subdirectory` with different paths violates that notion and any expected or intuitive understandings of the way that the CMake hierarchy works.

"Sparing the nasty details" can be accomplished through plenty of refactorings if this was significant or important. There could be more `*.cmake` modules, wrapped function modules, etc.

Really what you're suggesting with this `all/CMakeLists.txt` could be better divided into a structured like:

```
root/
  CMakeLists.txt -- This is your 'all'
  lib/CMakeList.txt -- This is your project
  test/CMakeLists.txt -- this is your test
```

Which makes it more intuitive from a topological perspective, all while achieving the same perceived benefits you are suggesting -- where now a user could simply include "lib" instead for just the library.

The proposed structure also misuses `project` calls, which in idiomatic CMake are meant to couple all components to a project (library, tests, any utilities, benchmarks, etc). This now has two `project` calls; one for the library, one for the "superbuild" which contains tests (and presumably any other utilities). 

--------------

At any rate, as I mentioned [above](https://github.com/bitwizeshift/result/pull/3#discussion_r593835292), this suggested refactoring appears more opinionated and subjective, and I haven't really seen anything yet that justifies such a change.

I do still appreciate you taking the time to bring this approach to my attention though! 🙂

Reply:
> project topology

The way I see it is that the project is technically a monorepo, whose packages are the library, tests, examples, docs, whatever else. When you have a monorepo, you ought to have a file the instruments the whole thing.  
You are correct in that monorepos normally should follow such a hierarchy, however this breaks assumptions by ExternalProject, FetchContent, CPM.cmake, vcpkg, which by default assume your root project CML to be at the root, not a superbuild CML. With these constraints, the normal definition of a monorepo has to be bent a little to fit our purposes. I believe I went into more details regarding this above already.

> I haven't really seen anything yet that justifies such a change

https://github.com/TheLartians/ModernCppStarter uses a similar structure and after thinking hard about it, it makes a lot of sense to me. This project is doing big numbers, along with CPM.cmake, so there is reason to think that there must be something in this.